### PR TITLE
feat: Accept cloud watch logs through amazon kinesis firehose

### PR DESCRIPTION
**Pull request by oasisk**

### CODE REVIEW

File: src/service/logs/kinesis_firehose.rs [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+                            Err(_e) => {
+                                local_val.insert("message".to_owned(), event.message.clone());
+                            }
+                            Ok(message_val) => {
+                                local_val.insert("message".to_owned(), message_val.clone());
+                            }
+                        }
+                    } else {
+                        local_val.insert("message".to_owned(), local_msg.into());
+                    }
```
Comment (vvv) [inline src/service/logs/kinesis_firehose.rs:150]:
```suggestion
                    {
                        let mut msg = event.message.as_str().unwrap();
                        if msg.starts_with('{') && msg.ends_with('}') {
                            msg = json::from_str::<json::Value>(local_msg).unwrap_or_else(|| event.essage.clone());
                        }
                        local_val.insert("message".to_owned(), msg);
                    }
```

File: src/service/logs/kinesis_firehose.rs [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        )
+        .await;
+    }
+    // Start get stream alerts
+    let key = format!("{}/{}/{}", &org_id, StreamType::Logs, &stream_name);
+    crate::service::ingestion::get_stream_alerts(key, &mut stream_alerts_map).await;
+    // End get stream alert
+
+    let mut buf: AHashMap<String, Vec<String>> = AHashMap::new();
+    for (_i, record) in request.records.iter().enumerate() {
```
Comment (vvv) [inline src/service/logs/kinesis_firehose.rs:113]: You don't use `_i`, so there's no need to enumerate.
```suggestion
    for record in request.records {
```

File: src/service/logs/kinesis_firehose.rs [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        match decode_and_decompress(&record.data) {
+            Ok(decompressed_data) => {
+                let kfh_data: KinesisFHData = json::from_str(&decompressed_data)?;
+
+                for event in kfh_data.log_events.iter() {
```
Comment (vvv) [inline src/service/logs/kinesis_firehose.rs:118]:
```suggestion
                for event in kfh_data.log_events {
```

@oasisk Did you check this code with clippy?

File: src/service/logs/kinesis_firehose.rs [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+                                stream_status.status.failed += 1;
+                                stream_status.status.error = e.to_string();
+                                continue;
+                            }
+                        },
+                        None => Utc::now().timestamp_micros(),
+                    };
+                    // check ingestion time
+                    let earliest_time =
+                        Utc::now() + Duration::hours(0 - CONFIG.limit.ingest_allowed_upto);
```
Comment (vvv) [inline src/service/logs/kinesis_firehose.rs:191]: This is peculiar.

Did you mean
```suggestion
                        Utc::now() - Duration::hours(CONFIG.limit.ingest_allowed_upto);
```
?

File: src/service/logs/kinesis_firehose.rs [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+                        request_id: request.request_id,
+                        error_message: Some(err.to_string()),
+                        timestamp: request.timestamp.unwrap_or(Utc::now().timestamp_micros()),
+                    }),
+                );
+            }
+        }
+    }
+
+    // write to file
```
Comment (vvv) [inline src/service/logs/kinesis_firehose.rs:237]: Please extract logically independent parts into separate functions.

File: src/service/logs/kinesis_firehose.rs [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    use super::decode_and_decompress;
+
+    #[test]
+    fn test_decode_and_decompress_success() {
+        let encoded_data = "H4sIAAAAAAAAADWO0QqCMBiFX2XsOkKJZHkXot5YQgpdhMTSPzfSTbaZhPjuzbTLj3M45xtxC1rTGvJPB9jHQXrOL2lyP4VZdoxDvMFyEKDmpJF9NVBTskTW2gaNrGMl+85mC2VGAW0X1P1Dl4p3hksR8caA0ti/Fb9e+AZhZhwxr5a64VbD0NaOuR5xPLJzycEh+81fbxa4JmjVQ6uejwIG5YuLGjGgjWFIPlFll7ig8zOKuAImNWzxVExfL8ipzewAAAA=";
+        let expected = "{\"messageType\":\"CONTROL_MESSAGE\",\"owner\":\"CloudwatchLogs\",\"logGroup\":\"\",\"logStream\":\"\",\"subscriptionFilters\":[],\"logEvents\":[{\"id\":\"\",\"timestamp\":1680683189085,\"message\":\"CWL CONTROL MESSAGE: Checking health of destination Firehose.\"}]}";
+        let result =
+            decode_and_decompress(encoded_data).expect("Failed to decode and decompress data");
+        assert_eq!(result, expected);
+    }
```
Comment (vvv) [inline src/service/logs/kinesis_firehose.rs:347]: [optional] Consider using [`expect-test`](https://docs.rs/expect-test/latest/expect_test/).

File: src/service/logs/kinesis_firehose.rs [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+
+    #[test]
+    fn test_decode_and_decompress_invalid_gzip() {
+        let encoded_data = "aGVsbG8gd29ybGQh"; // "hello world!" base64-encoded but not compressed
+        let result = decode_and_decompress(encoded_data);
+        assert!(
+            result.is_err(),
+            "Expected an error due to invalid gzip data"
+        );
```
Comment (vvv) [inline src/service/logs/kinesis_firehose.rs:366]: [nit]
```suggestion
        let invalid_input = "aGVsbG8gd29ybGQh"; // "hello world!" base64-encoded but not compressed
        assert!(decode_and_decompress(invalid_input).is_err());
```

**Review by vvv — APPROVED** [review]

File: src/main.rs [hunk 8]
```diff
@@ -183,11 +185,13 @@ async fn main() -> Result<(), anyhow::Error> {
...
                     .configure(get_basic_routes)
+                    .configure(get_other_service_routes)
             } else {
                 App::new().wrap(prometheus.clone()).service(
                     web::scope(&CONFIG.common.base_uri)
                         .configure(get_service_routes)
-                        .configure(get_basic_routes),
+                        .configure(get_basic_routes)
+                        .configure(get_other_service_routes),
                 )
```
Comment (hengfeiyang) [inline src/main.rs:195]: This part should keep baseic_routes at the end, the routes have priority.  basic one the top maybe cause problem.

**Review by hengfeiyang — APPROVED** [review]

File: src/handler/http/auth/mod.rs [hunk 9]
```diff
@@ -133,6 +134,41 @@ pub async fn validate_user(user_id: &str, user_password: &str) -> Result<bool, E
     }
 }
 
+pub async fn validator_amz(
```
Comment (hengfeiyang) [inline src/handler/http/auth/mod.rs:137]: Maybe change the function name to `validator_aws`?

**Pull request merged**